cri: Don't use rel path for image volumes

rata opened the pull request:
Runc 1.1 throws a warning when using rel destination paths, and runc 1.2 is planning to thow an error (i.e. won't start the container).

Let's just make this an abs path in the only place it might not be: the mounts created due to `VOLUME` directives in the Dockerfile.

---

We discussed this issue in today's containerd community meeting and decided to make this small PR that should fix it, and if this is merged without concerns, backport it to 1.6 and 1.7.

cc @mikebrow

mikebrow commented on pkg/cri/sbserver/container_create.go:
```diff
@@ -355,6 +356,9 @@ func (c *criService) volumeMounts(containerRootDir string, criMounts []*runtime.
 		src := filepath.Join(containerRootDir, "volumes", volumeID)
+		if !filepath.IsAbs(dst) && goruntime.GOOS != "windows" {
+			dst = filepath.Join("/", dst)
+		}
```
maybe a func for this when not goos windows.. something like https://github.com/containerd/containerd/blob/main/remotes/docker/config/hosts.go#L535-L540 

and perhaps a debug out when joining the base to the dst.. should be a very rare occurrence..

rata replied:
Following @gabriel-samfira suggestions, the code ended up being different in sbserver and server. So I don't think a function is worth now.

Also, for logging, I need to pass the context to volumeMounts, which doesn't seem worth it, IMHO.

Please re-open if you think it is :)

mikebrow replied:
see below comment...

mikebrow reviewed (commented):
see suggestion..

tagging related PR https://github.com/containerd/containerd/pull/8524 @gabriel-samfira

mikebrow commented:
As discussed in the community meeting, perhaps a switch is nec. in runc, deprecated/temporary, to enable ignoring/disabling the abs check, until such time as the affected container runtimes have been patched and deployed and/or are no longer in support.

mikebrow commented:
/ok-to-test

gabriel-samfira commented on pkg/cri/sbserver/container_create.go:
```diff
@@ -355,6 +356,9 @@ func (c *criService) volumeMounts(containerRootDir string, criMounts []*runtime.
 		}
 		volumeID := util.GenerateID()
 		src := filepath.Join(containerRootDir, "volumes", volumeID)
+		if !filepath.IsAbs(dst) && goruntime.GOOS != "windows" {
```
This will most likely break any runtime that is capable of running containers for an OS different that the one that `containerd` is running on. For example, Windows is capable of running both Windows containers (WCOW) and Linux containers (LCOW). For Linux containers, `/C:/test` is a valid path. For Windows containers, it is not. Other runtimes may be written in the future,  that can run Windows containers on Linux. Any runtime that leverage virtual machines are potentially capable of this.

Paths should be handled based on the sandbox target `OS`, not on `runtime.GOOS`.

The  `volumeMounts()` function ends up being [called in `CreateContainer()`](https://github.com/containerd/containerd/blob/main/pkg/cri/sbserver/container_create.go#L154) which passes the mounts returned by this function to `buildContainerSpec()`. In `buildContainerSpec()` [we check `platform.OS`](https://github.com/containerd/containerd/blob/main/pkg/cri/sbserver/container_create.go#L522-L573) before building the spec.

This would be a better place to check the volume `dst` path and decide to normalize it if the target OS is `Unix`.

fuweid replied:
Good point. It will be correct to add the platform as argument in `volumeMounts` for `sbserver` package.

rata replied:
Ohh, great points. Added the `platform` param to sbserver as @fuweid. Thanks!

gabriel-samfira commented on pkg/cri/server/container_create.go:
```diff
@@ -332,6 +333,9 @@ func (c *criService) volumeMounts(containerRootDir string, criMounts []*runtime.
 		}
 		volumeID := util.GenerateID()
 		src := filepath.Join(containerRootDir, "volumes", volumeID)
+		if !filepath.IsAbs(dst) && goruntime.GOOS != "windows" {
```
The volumes returned by this function, end up being passed to [`containerSpec()`](https://github.com/containerd/containerd/blob/main/pkg/cri/server/container_create.go#L159), which is platform specific. The better place to handle the path would be closer to where it actually ends up being used. This way each OS can handle it in accordance to the rules of that OS.

fuweid replied:
Since the `cri/server` doesn't use sandbox API, we don't know the runtime platform actually, if I don't miss something here. So, I think we can keep it as it is here. Use host platform as condition. And it's easy to backport and maintain in LTS.

gabriel-samfira replied:
That's true. My thought here was that since this ends up being used in platform specific files [`container_create_linux.go`](https://github.com/containerd/containerd/blob/main/pkg/cri/server/container_create_linux.go#L117), [`container_create_windows.go`](https://github.com/containerd/containerd/blob/main/pkg/cri/server/container_create_windows.go#L41), etc, it would spare us the extra check on `runtime.GOOS`.

Either way is fine with me.

rata replied:
I'll let it like this as @fuweid and Derek also both said they prefer to just change it here, rather than in those functions that can affect mounts created by other means.

Thanks!

mikebrow replied:
nod I wasn't pointing over to the mount stuff to say call that I meant a new one local to both observer and server.. 

platform is just a type alias for oci platform.. and from what I see so far pod sandbox is ...
```
func (c *Controller) Platform(_ctx context.Context, _sandboxID string) (platforms.Platform, error) {
	return platforms.DefaultSpec(), nil
}
func DefaultSpec() specs.Platform {
	return specs.Platform{
		OS:           runtime.GOOS,
		Architecture: runtime.GOARCH,
		// The Variant field will be empty if arch != ARM.
		Variant: cpuVariant(),
	}
} 
```
you could just pass platform.OS from sbserver and pass runtime.GOOS from the other for now.. if we had a common util func.. The only reason really to suggest common was in case we need to use that common in other locations in server/sbserver code and not have to remember this pattern again :-)

gabriel-samfira reviewed (commented):
[Handling paths is tricky](https://github.com/moby/buildkit/pull/3322#issuecomment-1479812417). Generally speaking, paths [should be "normalized" based on target OS](https://github.com/moby/buildkit/blob/master/util/system/path.go). The target OS in `containerd` is generally set by the sandbox image.

In the case of this PR, I think we can get the desired result if we move the check for relative paths closer to where those volumes need used.

rata commented on integration/volume_copy_up_test.go:
```diff
@@ -93,7 +93,7 @@ func TestVolumeCopyUp(t *testing.T) {
 			},
 		},
 		{
-			containerPath: "C:/weird_test_dir",
```
Right, but we are only testing that non-linux does the right thing, as when we run the test on windows, we just overwrite the expected volumes with: https://github.com/containerd/containerd/blob/81895d22c9ee091db6f9371e3177b50744fe2edb/integration/volume_copy_up_test.go#L106-L122

gabriel-samfira replied:
Ahh! Yes. You are right. I missed that part.

rata replied:
Cool, I think this is solved now. Let me know if I'm missing something and re-open :)

fuweid reviewed (approved):
LGTM

gabriel-samfira reviewed (commented):
LGTM. Some tests seem to be failing, but otherwise seems ok.

rata commented:
@gabriel-samfira @fuweid thanks! Re-pushed so tests run again, it didn't seem related to this PR.

rata commented:
@gabriel-samfira @fuweid fixed the CI now, PTAL (the tests I added were quite useful!).

The issue was that when executing on Windows, filepath.Join() does a clean, which changes from slash to backslash, and didn't match the expected paths.

I've adjusted the code to work fine on windows and linux, and added a log line as @mikebrow suggested. The changes were basically [this](https://github.com/containerd/containerd/pull/8885/files#diff-15877ae3602deacafa14a8688676328a93c85215593babe9f4b8a7bd46c2803bR359-R368).

mikebrow reviewed (approved):
LGTM

mikebrow commented on pkg/cri/sbserver/container_create.go:
```diff
@@ -355,6 +356,16 @@ func (c *criService) volumeMounts(containerRootDir string, criMounts []*runtime.
+			// On Linux, ToSlash() is a no-op.
+			oldDst := dst
+			dst = filepath.ToSlash(filepath.Join("/", dst))
+			log.L.Infof("Volume destination %q is not absolute, converted to %q", oldDst, dst)
```
debug here.. info to noisy

rata replied:
Sure, changed to debug.

mikebrow commented on pkg/cri/server/container_create.go:
```diff
@@ -332,6 +333,11 @@ func (c *criService) volumeMounts(containerRootDir string, criMounts []*runtime.
+		if !filepath.IsAbs(dst) && goruntime.GOOS != "windows" {
+			oldDst := dst
+			dst = filepath.Join("/", dst)
+			log.L.Infof("Volume destination %q is not absolute, converted to %q", oldDst, dst)
```
debug pls

rata replied:
I kept log.L, as we don't have a context here, but changed it to Debugf

fuweid commented:
@rata would you update with mike's suggestion? after update, I will merge it. Thanks

rata commented:
@fuweid just changed the log line from info to debug, thanks!

fuweid reviewed (approved):
LGTM

Thompson1985 reviewed (commented):
Reviewed
